src/main/java/me/chrr/scribble/tool/commandmanager/CommandManager.java

```diff
@@ -0,0 +1,130 @@
+
+    public CommandManager(int maxHistorySize) {
+        this.maxHistorySize = maxHistorySize;
+        this.commandStack = new LinkedList<>();
+    }
+
+    /**
+     * Clears the command history and resets the manager's state.
+     * All previously executed commands are forgotten.
+     */
+    public void clear() {
+        lastExecutedCommandIndex = EMPTY_STACK_INDEX;
+        commandStack.clear();
+    }
+
+    /**
+     * Executes the given command and clears any commands that were previously undone.
+     * If the history stack exceeds the maximum size, the oldest command is removed.
+     * <p>
+     * The command won't be added to the history stack if it executed with result value false
+     *
+     * @param command The command to execute.
+     */
+    public void execute(@NotNull Command command) {
+        // dropAllAboveCurrentIndex is needed for cases like:
+        // execute A - currentIndex at A
+        // execute B - currentIndex at B
+        // execute C - currentIndex at C
+        // undo - moves currentIndex to B
+        // undo - moves currentIndex to A
+        // execute D - will drop all commands that were above currentIndex(above A)
+        dropAllAboveCurrentIndex();
+
+        if (commandStack.size() >= maxHistorySize && maxHistorySize > 0) {
+            // size limit was reached
+            // drop a command from the bottom of the stack
+            commandStack.pollFirst();
+            // also more the current index by one
+            lastExecutedCommandIndex--;
+        }
+
+        boolean affectedState = command.execute();
+        if (affectedState && maxHistorySize > 0) {
+            commandStack.add(command);
+            lastExecutedCommandIndex++;
+        }
+    }
+
+    /**
+     * Removes any commands that are above the current execution point in the history stack.
+     */
+    private void dropAllAboveCurrentIndex() {
+        while (lastExecutedCommandIndex < commandStack.size() - 1) {
+            commandStack.pollLast();
+        }
+    }
+
+    /**
+     * Attempts to undo the last executed commands,
+     * until one of commands undo is successful or has nothing to undo.
+     *
+     * @return True if the undo operation was successful, false otherwise.
+     */
+    public boolean tryUndo() {
+        if (hasCommandsToUndo()) {
+            boolean wasRolledBack = commandStack.get(lastExecutedCommandIndex).rollback();
+            lastExecutedCommandIndex--;
+
+            if (!wasRolledBack) {
+                // rolling back until the rollback is successful
+                return tryUndo();
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks if an undo operation is possible.
+     *
+     * @return True if an undo operation is possible, false otherwise.
+     */
+    public boolean hasCommandsToUndo() {
+        if (commandStack.isEmpty()) {
+            lastExecutedCommandIndex = EMPTY_STACK_INDEX;
```

Comment:
This one feels weird. Generally state should not change in a getter, and I think there's better places to put this.

Reply:
I agree. It's completely unnecessary here, I left it just as safe measure during development. I'll remove it.